test/unit/ReservoirRouter.t.sol

```diff
@@ -0,0 +1,759 @@
+        _tokenA.mint(address(_constantProductPair), lAmountIn);
+        uint256 lSwapFee = _constantProductPair.swapFee();
+
+        // act
+        uint256 lAmountOut = _router.getAmountOut(lAmountIn, lReserve0, lReserve1, 0, lSwapFee, ExtraData(0,0,0));
+        uint256 lActualAmountOut = _constantProductPair.swap(int256(lAmountIn), true, address(this), bytes(""));
+
+        // assert
+        assertLt(lAmountOut, lAmountIn);
+        assertEq(lAmountOut, lActualAmountOut);
+    }
+
+    function testGetAmountOut_SP(uint256 aAmountIn) public
+    {
+        // assume
+        uint256 lAmountIn = bound(aAmountIn, 1, type(uint112).max);
+
+        // arrange
+        (uint112 lReserve0, uint112 lReserve1, ) = _stablePair.getReserves();
+        _tokenA.mint(address(_stablePair), lAmountIn);
+        uint256 lSwapFee = _stablePair.swapFee();
+        uint64 lToken0PrecisionMultiplier = ReservoirLibrary.getPrecisionMultiplier(_stablePair.token0());
+        uint64 lToken1PrecisionMultiplier = ReservoirLibrary.getPrecisionMultiplier(_stablePair.token1());
+        uint64 lA = ReservoirLibrary.getAmplificationCoefficient(address(_stablePair));
+
+        // act
+        uint256 lAmountOut
+            = _router.getAmountOut(
+                lAmountIn,
+                lReserve0,
+                lReserve1,
+                1,
+                lSwapFee,
+                ExtraData(lToken0PrecisionMultiplier,lToken1PrecisionMultiplier, lA)
+        );
+        uint256 lActualAmountOut = _stablePair.swap(int256(lAmountIn), true, address(this), bytes(""));
+
+        // assert
+        assertLt(lAmountOut, lAmountIn);
+        assertEq(lAmountOut, lActualAmountOut);
+    }
+
+    function testGetAmountIn_ErrorChecking(uint256 aCurveId, uint256 aAmountOut) public
+    {
+        // assume
+        aCurveId = bound(aCurveId, 0, 1);
+        uint256 aAmountOut = bound(aAmountOut, 1, type(uint112).max);
+
+        // act & revert
+        vm.expectRevert("RL: INSUFFICIENT_OUTPUT_AMOUNT");
+        _router.getAmountIn(0, 10, 10, aCurveId, 30, ExtraData(0,0,0));
+
+        vm.expectRevert("RL: INSUFFICIENT_LIQUIDITY");
+        _router.getAmountIn(aAmountOut, 0, 0, aCurveId, 0, ExtraData(0,0,0));
+    }
+
+    function testGetAmountIn_CP(uint256 aAmountOut) public
+    {
+        // assume
+        (uint112 lReserve0, uint112 lReserve1, ) = _constantProductPair.getReserves();
+        uint256 lAmountOut = bound(aAmountOut, 1000, lReserve1 / 2);
+
+        // arrange
+        uint256 lSwapFee = _constantProductPair.swapFee();
+
+        // act
+        uint256 lAmountIn = _router.getAmountIn(lAmountOut, lReserve0, lReserve1, 0, lSwapFee, ExtraData(0,0,0));
+        _tokenA.mint(address(_constantProductPair), lAmountIn);
+        uint256 lActualAmountOut = _constantProductPair.swap(-int256(lAmountOut), false, address(this), bytes(""));
+
+        // assert
+        assertLt(lAmountOut, lAmountIn);
+        assertEq(lAmountOut, lActualAmountOut);
```

Comment:
I would propose moving all the view tests into a `ReservoirLibrary.t.sol`, then here you would just differential fuzz the router against the `ReservoirLibrary` implementation (proving they are the same). Then inside if the ReservoirLibrary test we can write a bunch of manunal/semi-manual mathematical checks.

Additionally, inside of the router tests I would add tests that make sure the estimate is always the same as the output (e.g. fuzz random trades and then check whether the output matches for both quote and execute).

Reply:
okay moved them into a separate test lib. 

Also introduced 2 differential tests in router test. If this approach is correct I will add the two `getAmountIn` test. I think it's not necessary to differential test `getAmountsOut` and `getAmountsIn` since the router simply calls the lib and both functions are based on the two more basic functions which are already being differential tested.

In the router swap tests I'm already doing the check of quote vs execute. So we're good on that front